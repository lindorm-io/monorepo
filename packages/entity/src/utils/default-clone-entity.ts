import { isFunction } from "@lindorm/is";
import { Constructor } from "@lindorm/types";
import { IEntity } from "../interfaces";
import { MetaColumnDecorator } from "../types";
import { globalEntityMetadata } from "./global";
import { parseColumn } from "./private";

const reset: Array<MetaColumnDecorator> = [
  "CreateDateColumn",
  "ExpiryDateColumn",
  "DeleteDateColumn",
  "UpdateDateColumn",
  "VersionColumn",
];

export const defaultCloneEntity = <E extends IEntity>(
  Entity: Constructor<E>,
  entity: E,
): E => {
  const metadata = globalEntityMetadata.get(Entity);
  const clone = new Entity();

  for (const column of metadata.columns) {
    if (metadata.generated.some((g) => g.key === column.key)) continue;

    if (
      reset.includes(column.decorator) ||
      metadata.primaryKeys.includes(column.key) ||
      metadata.generated.some((g) => g.key === column.key) ||
      metadata.indexes.some((u) => u.unique && u.keys.every((k) => k.key === column.key))
    ) {
      (clone as any)[column.key] = undefined;
    } else {
      (clone as any)[column.key] = parseColumn(column, entity);
    }

    if ((clone as any)[column.key]) continue;

    (clone as any)[column.key] = isFunction(column.fallback)
      ? column.fallback()
      : column.fallback;
  }

  const hooks = metadata.hooks.filter((h) => h.decorator === "OnCreate");

  for (const hook of hooks) {
    hook.callback(entity);
  }

  return clone;
};
