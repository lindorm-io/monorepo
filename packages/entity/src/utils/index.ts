export * from "./default-clone-entity";
export * from "./default-create-entity";
export * from "./default-generate-entity";
export * from "./default-update-entity";
export * from "./default-validate-entity";
export * from "./get-collection-name";
export * from "./get-increment-name";
export * from "./get-save-strategy";
export * from "./global";
export * from "./remove-readonly";
export * from "./verify-readonly";
