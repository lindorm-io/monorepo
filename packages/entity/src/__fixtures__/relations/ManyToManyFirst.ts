import { Entity, ManyToMany, PrimaryKeyColumn } from "../../decorators";
import { ManyToManySecondDecoratorEntity } from "./ManyToManySecond";

@Entity()
export class ManyToManyFirstDecoratorEntity {
  @PrimaryKeyColumn()
  id!: string;

  @ManyToMany(() => ManyToManySecondDecoratorEntity, "first", true)
  second!: Array<ManyToManySecondDecoratorEntity>;
}
