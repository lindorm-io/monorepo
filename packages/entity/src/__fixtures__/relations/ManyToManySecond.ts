import { Entity, ManyToMany, PrimaryKeyColumn } from "../../decorators";
import { ManyToManyFirstDecoratorEntity } from "./ManyToManyFirst";

@Entity()
export class ManyToManySecondDecoratorEntity {
  @PrimaryKeyColumn()
  id!: string;

  @ManyToMany(() => ManyToManyFirstDecoratorEntity, "second")
  first!: Array<ManyToManyFirstDecoratorEntity>;
}
