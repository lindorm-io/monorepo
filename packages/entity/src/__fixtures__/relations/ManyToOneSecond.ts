import { Entity, ManyToOne, PrimaryKeyColumn } from "../../decorators";
import { OneToManyFirstDecoratorEntity } from "./OneToManyFirst";

@Entity()
export class ManyToOneSecondDecoratorEntity {
  @PrimaryKeyColumn()
  id!: string;

  @ManyToOne(() => OneToManyFirstDecoratorEntity, "second")
  first!: OneToManyFirstDecoratorEntity;
}
