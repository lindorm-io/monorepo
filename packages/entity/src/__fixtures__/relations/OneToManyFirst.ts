import { Entity, OneToMany, PrimaryKeyColumn } from "../../decorators";
import { ManyToOneSecondDecoratorEntity } from "./ManyToOneSecond";

@Entity()
export class OneToManyFirstDecoratorEntity {
  @PrimaryKeyColumn()
  id!: string;

  @OneToMany(() => ManyToOneSecondDecoratorEntity, "first")
  second!: Array<ManyToOneSecondDecoratorEntity>;
}
