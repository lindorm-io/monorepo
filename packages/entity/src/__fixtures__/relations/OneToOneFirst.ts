import { Entity, OneToOne, PrimaryKeyColumn } from "../../decorators";
import { OneToOneSecondDecoratorEntity } from "./OneToOneSecond";

@Entity()
export class OneToOneFirstDecoratorEntity {
  @PrimaryKeyColumn()
  id!: string;

  @OneToOne(() => OneToOneSecondDecoratorEntity, "first", true)
  second!: OneToOneSecondDecoratorEntity;
}
