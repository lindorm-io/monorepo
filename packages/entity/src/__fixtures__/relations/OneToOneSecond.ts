import { Entity, OneToOne, PrimaryKeyColumn } from "../../decorators";
import { OneToOneFirstDecoratorEntity } from "./OneToOneFirst";

@Entity()
export class OneToOneSecondDecoratorEntity {
  @PrimaryKeyColumn()
  id!: string;

  @OneToOne(() => OneToOneFirstDecoratorEntity, "second", false)
  first!: OneToOneFirstDecoratorEntity;
}
