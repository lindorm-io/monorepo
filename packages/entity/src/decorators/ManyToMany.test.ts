import { ManyToManyFirstDecoratorEntity } from "../__fixtures__/relations/ManyToManyFirst";
import { ManyToManySecondDecoratorEntity } from "../__fixtures__/relations/ManyToManySecond";
import { globalEntityMetadata } from "../utils";

describe("ManyToMany Decorator", () => {
  test("should add metadata", () => {
    expect(globalEntityMetadata.get(ManyToManyFirstDecoratorEntity)).toMatchSnapshot();
    expect(globalEntityMetadata.get(ManyToManySecondDecoratorEntity)).toMatchSnapshot();

    const Second = globalEntityMetadata
      .get(ManyToManyFirstDecoratorEntity)
      .relations[0]!.foreignConstructor();
    const First = globalEntityMetadata
      .get(ManyToManySecondDecoratorEntity)
      .relations[0]!.foreignConstructor();

    expect(new Second()).toBeInstanceOf(ManyToManySecondDecoratorEntity);
    expect(new First()).toBeInstanceOf(ManyToManyFirstDecoratorEntity);
  });
});
