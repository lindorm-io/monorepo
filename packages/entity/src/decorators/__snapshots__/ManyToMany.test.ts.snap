// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManyToMany Decorator should add metadata 1`] = `
{
  "columns": [
    {
      "decorator": "PrimaryKeyColumn",
      "enum": null,
      "fallback": [Function],
      "key": "id",
      "max": null,
      "min": null,
      "nullable": false,
      "optional": false,
      "readonly": true,
      "schema": null,
      "type": "uuid",
    },
  ],
  "entity": {
    "cache": null,
    "database": null,
    "decorator": "Entity",
    "name": null,
    "namespace": null,
  },
  "extras": [],
  "generated": [],
  "hooks": [],
  "indexes": [],
  "primaryKeys": [
    "id",
  ],
  "primarySource": null,
  "relations": [
    {
      "foreignConstructor": [Function],
      "foreignKey": "first",
      "key": "second",
      "options": {
        "joinKey": null,
        "joinTable": true,
        "loading": null,
        "nullable": false,
        "onDelete": null,
        "onOrphan": null,
        "onUpdate": null,
        "strategy": null,
      },
      "type": "ManyToMany",
    },
  ],
  "schemas": [],
}
`;

exports[`ManyToMany Decorator should add metadata 2`] = `
{
  "columns": [
    {
      "decorator": "PrimaryKeyColumn",
      "enum": null,
      "fallback": [Function],
      "key": "id",
      "max": null,
      "min": null,
      "nullable": false,
      "optional": false,
      "readonly": true,
      "schema": null,
      "type": "uuid",
    },
  ],
  "entity": {
    "cache": null,
    "database": null,
    "decorator": "Entity",
    "name": null,
    "namespace": null,
  },
  "extras": [],
  "generated": [],
  "hooks": [],
  "indexes": [],
  "primaryKeys": [
    "id",
  ],
  "primarySource": null,
  "relations": [
    {
      "foreignConstructor": [Function],
      "foreignKey": "second",
      "key": "first",
      "options": {
        "joinKey": null,
        "joinTable": false,
        "loading": null,
        "nullable": false,
        "onDelete": null,
        "onOrphan": null,
        "onUpdate": null,
        "strategy": null,
      },
      "type": "ManyToMany",
    },
  ],
  "schemas": [],
}
`;
